### PR TITLE
Add more DECFLOAT to C type conversions

### PR DESCRIPTION
**Pull request by sfc-gh-pbulawa**
### TL;DR

Added comprehensive ODBC type conversion support for DECFLOAT to all standard C data types including integers, floating-point, character, binary, numeric, bit, and interval types.

### What changed?

Extended the `SnowflakeDecfloat` implementation to support conversion to all ODBC C data types:

- **Integer types**: Added support for `SQL_C_LONG`, `SQL_C_SHORT`, `SQL_C_TINYINT`, `SQL_C_SBIGINT`, `SQL_C_UBIGINT` and their signed/unsigned variants with proper range checking and fractional truncation warnings
- **Floating-point types**: Implemented `SQL_C_DOUBLE` and `SQL_C_FLOAT` conversions with overflow detection
- **Bit type**: Added `SQL_C_BIT` support with validation for 0/1 values only
- **Numeric type**: Enhanced `SQL_C_NUMERIC` conversion with proper precision and scale handling
- **Binary type**: Implemented `SQL_C_BINARY` conversion that outputs `SQL_NUMERIC_STRUCT` format
- **Interval types**: Added support for single-field interval types (`SQL_C_INTERVAL_YEAR`, `SQL_C_INTERVAL_MONTH`, etc.) with proper precision validation and fractional handling
- **Helper functions**: Added `compute_int_and_fractional()`, `decfloat_to_f64()`, `check_integer_range()`, and `fractional_warning()` utilities
- **Error handling**: Integrated `NumericValueOutOfRangeSnafu` and `IntervalFieldOverflowSnafu` for overflow conditions and proper warning generation for fractional truncation

Replaced `abs_sig % 10 == 0` with `abs_sig.is_multiple_of(10)` for better readability in the formatting function.

### How to test?

Run the comprehensive test suite that includes:

- Unit tests in `decfloat_tests.rs` covering all conversion scenarios including extreme exponents, overflow conditions, and edge cases
- End-to-end C++ tests for each target type with boundary conditions, overflow cases, fractional truncation, and NULL handling
- Feature files defining test scenarios for all supported conversions including interval types with custom precision handling
- Tests for incompatible conversions (temporal and GUID types) that should return appropriate error codes
- Parameter binding tests that now work with server-side conversion

### Why make this change?

This change provides complete ODBC specification compliance for DECFLOAT type conversions, enabling applications to retrieve DECFLOAT values in any supported C data type format. The implementation ensures proper error handling, range validation, and warning generation according to ODBC standards, making the driver more robust and compatible with diverse client applications. It also enables parameter binding for DECFLOAT values, removing previous limitations.

### CODE REVIEW

**Comment by sfc-gh-pbulawa** [comment]
* **#734** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/734?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/734?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#728** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/728?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: odbc/src/conversion/decfloat.rs [hunk 1]
```diff
@@ -166,6 +167,54 @@ pub(crate) fn format_decfloat(sig: i128, exp: i16, max_plain_digits: usize) -> S
...
+            None => (0, sig != 0),
+        }
+    }
+}
+
+fn decfloat_to_f64(sig: i128, exp: i16) -> f64 {
+    (sig as f64) * 10f64.powi(exp as i32)
+}
+
+fn check_integer_range(value: i128, min: i128, max: i128) -> Result<(), WriteOdbcError> {
```
Comment (sfc-gh-makowalski) [inline odbc/src/conversion/decfloat.rs]: We have the same exact function in `number.rs`:L122

File: odbc/src/conversion/decfloat.rs [hunk 2]
```diff
@@ -166,6 +167,54 @@ pub(crate) fn format_decfloat(sig: i128, exp: i16, max_plain_digits: usize) -> S
...
+        NumericValueOutOfRangeSnafu {
+            reason: format!("Value {value} is out of range ({min} to {max})"),
+        }
+        .fail()
+    } else {
+        Ok(())
+    }
+}
+
+fn fractional_warning(has_fractional: bool) -> Warnings {
```
Comment (sfc-gh-makowalski) [inline odbc/src/conversion/decfloat.rs]: We have the same exact function in `number.rs`:L133

File: odbc/src/conversion/decfloat.rs [hunk 3]
```diff
@@ -200,7 +249,282 @@ impl WriteODBCType for SnowflakeDecfloat {
...
+                                numeric_bytes.as_ptr(),
+                                binding.target_value_ptr as *mut u8,
+                                numeric_size,
+                            );
+                        }
+                        let _ = binding.write_length_or_null(
+                            crate::conversion::traits::LengthOrNull::Length(
+                                numeric_size as sql::Len,
+                            ),
+                        );
```
Comment (sfc-gh-makowalski) [inline odbc/src/conversion/decfloat.rs]: We have the exact same snippet in `real.rs`:L287

**Review by sfc-gh-makowalski — APPROVED** [review]
The PR overall looks good, not sure if we'd like to create helpers for the duplicated code parts that I pointed out. I'll leave it up to you to decide. Approving in advance.

**Review by sfc-gh-makowalski — APPROVED** [review]

**Review by sfc-gh-makowalski — APPROVED** [review]
